daml-lf/spec/daml-lf-1.rst

```diff
@@ -1854,6 +1856,95 @@ will always be used to compare values of same types::
 .. note:: the equality of generic map is not sensitive to the order of
           its entries. See rules ``'GenEqNonEmptyGenMap'``.
 
+Value order
+~~~~~~~~~~~
+
+In this section, we define a strict partial order relation ``<ᵥ`` on values.
+This is a strict order when comparing serialized values of the same type.
+
+We also define the transitive relation ``≲ᵥ`` as the union of ``~ᵥ`` and
+``<ᵥ``. This relation is transitive, and antisymmetric with respect to ``~ᵥ``.
+It is a total order when comparing serialized values of the same type.
+
+
+                          ┌────────┐
+  Value Order Relation    │ v <ᵥ w │
+                          └────────┘
+
+  ——————————————————————————————————————————————————— GenLtTrueFalse
+   'False' <ᵥ 'True'
+
+   LitNumeric₁ is less than LitNumeric₂ as numbers.
+  ——————————————————————————————————————————————————— GenLtLitNumeric
+   LitNumeric₁ <ᵥ LitNumeric₂
+
+   t₁ comes lexicographically strictly before t₂,
+   when viewed as sequences of Unicode code points
+  ——————————————————————————————————————————————————— GenLtLitText
+   t₁ <ᵥ t₂
+
+   LitDate₁ is strictly before LitDate₂ as dates
+  ——————————————————————————————————————————————————— GenLtLitDate
+   LitDate₁ <ᵥ LitDate₂
+
+   LitTimestamp₁ is strictly before LitTimestamp₂ as
+   timestamps
+  ——————————————————————————————————————————————————— GenLtLitTimestamp
+   LitTimestamp₁ <ᵥ LitTimestamp₂
+
+   LitParty₁ comes lexicographically before
+   LitParty₂ when viewed as sequences of Unicode
+   code points
+  ——————————————————————————————————————————————————— GenLtLitParty
+   LitParty₁ <ᵥ LitParty₂
+
+   cid₁ is ordered strictly before cid₂ according
+   to the ledger's rules
+  ——————————————————————————————————————————————————— GenLtLitContractId
+   cid₁ <ᵥ cid₂
+
+  ——————————————————————————————————————————————————— GenLtListNil
+   'Nil' @τ <ᵥ 'Cons' @σ wₜ wₜ
+
+   vₕ <ᵥ wₕ
+  ——————————————————————————————————————————————————— GenLtListConsHead
+   'Cons' @τ vₕ vₜ  <ᵥ 'Cons' @σ wₜ wₜ
+
+   vₕ ~ᵥ wₕ    vₜ <ᵥ wₜ
+  ——————————————————————————————————————————————————— GenLtListConsTail
+   'Cons' @τ vₕ vₜ  <ᵥ 'Cons' @σ wₜ wₜ
+
+  ——————————————————————————————————————————————————— GenLtOptionalNone
+   'None' @τ <ᵥ 'Some' @σ w
+
+   v <ᵥ w
+  ——————————————————————————————————————————————————— GenLtOptionalSome
+   'Some' @τ v <ᵥ 'Some' @σ w
+
+   v₁ ~ᵥ v₁   ⋯   vᵢ₋₁ ~ᵥ wᵢ₋₁     vᵢ <ᵥ wᵢ    i <= n
+  ——————————————————————————————————————————————————— GenLtRecCon
+  Mod:T @τ1 … @τₙ { f₁ = v₁, …, fₙ = wₘ }
+    <ᵥ Mod:T @σ₁ … @σₙ { f₁ = w₁, …, fₙ = wₘ }
+
+   Mod:T:V₁ comes before Mod:T:V₂ in the list of
+   constructors for variant type Mod:T
```

Comment:
This is currently tricky to do in the engine as we do not have enough info at run time.
I was thinking to go lexicographically, but this is definitively better.

Reply:
I don’t think lexicographic ordering is a reasonable option. The default `Ord` instance in DAML works like the one that @associahedron has used in the spec here and diverging from that is going to cause a lot of confusion.

Reply:
We probably add an index in the variant and the enum.